lib/internal/util/comparisons.js

```diff
@@ -78,22 +101,37 @@ function strictDeepEqual(val1, val2, memos) {
   if (val1Tag !== val2Tag) {
     return false;
   }
-  if (Object.getPrototypeOf(val1) !== Object.getPrototypeOf(val2)) {
+  if (getPrototypeOf(val1) !== getPrototypeOf(val2)) {
     return false;
   }
   if (val1Tag === '[object Array]') {
     // Check for sparse arrays and general fast path
-    if (val1.length !== val2.length)
+    if (val1.length !== val2.length) {
+      return false;
+    }
+    const keys = objectKeys(val1);
+    if (keys.length !== objectKeys(val2).length) {
       return false;
-    // Skip testing the part below and continue with the keyCheck.
-    return keyCheck(val1, val2, true, memos);
+    }
+    // Fast path for non sparse arrays (no key comparison for indices
```

Comment:
How does this `if` check assert this is a fast path and the array is not sparse?

Reply:
From what I can tell:
 - this just checks that all the keys are indexed properties and that the last key is the last index
 - it relies on the `Object.keys` key ordering in V8 - AFAIK there is no guarantee on keys but there is on the order of other stuff e.g. getOwnPropertyNames (I am fine with this and just tracking it in case it changes in V8).
 - It fails with setting negative indices, for example: `var o = {}; o[-1] = 3; o[1] = 5; o.length = 2`. We should add a test for this particular case. For what it's worth I'm fine with that case being slow since it's uncommon and sparse arrays are really confusing perf-wise anyway.

Reply:
Alternatively, we can see if we can check it directly from V8 - something like `UseSparseVariant` - or use `%HasComplexElements`

Reply:
First, the keys that are integer indices (what these are is explained later), in ascending numeric order.
Then, all other string keys, in the order in which they were added to the object.
Lastly, all symbol keys, in the order in which they were added to the object.

Reply:
The keys are ordered by [spec](https://tc39.github.io/ecma262/#sec-ordinaryownpropertykeys) as follows:

1. The keys that are integer indices, in ascending numeric order.
2. All string keys, in the order in which they were added to the object.
3. All symbol keys, in the order in which they were added to the object.

The order makes this possible: `Object.keys(arr)[arr.length - 1] === '' + (arr.length - 1)`. By checking the key at `arr.length - 1` it should be the same key as the key length. Otherwise the key would have a position lower than that. If `Object.keys(arr).length === arr.length` on top of that, we also know that the array has no extra keys besides potentially symbols.

Reply:
@BridgeAR as I said above as far as I know `OrdinaryOwnPropertyKeys` is used in `.getOwnPropertyNames` but not in `.keys` - that said I don't particularly mind it since I don't see a future in which V8 changes the iteration order of objects anyway since it's so heavily relied on.

I think negative indices are not a problem since they're not technically indices from the spec PoV and would be enumerated elsewhere. 

Upon further inspection - it looks like this works fine for all the cases - care to add a comment?

Reply:
By spec [`Object.keys`](https://tc39.github.io/ecma262/#sec-object.keys) calls [`EnumerableOwnPropertyNames`](https://tc39.github.io/ecma262/#sec-enumerableownpropertynames) and that calls [`OwnPropertyKeys`](https://tc39.github.io/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-ownpropertykeys) and that calls [`OdinaryOwnPropertyKeys`](https://tc39.github.io/ecma262/#sec-ordinaryownpropertykeys) :)

Reply:
Neat, I wonder when it changed - I remember this discussion when we did the io changes for ES2015 and it was guaranteed for one and not the other. Good to know it's a guarantee now - thanks.